c2rust-analyze/src/known_fn.rs

```diff
@@ -602,28 +622,823 @@ pub const fn all_known_fns() -> &'static [KnownFn] {
             ) -> c_int;
 
             fn access(
-                path: *const c_char: [READ | OFFSET_ADD],
+                path: *const c_char: [READ | OFFSET_ADD | NON_NULL],
                 amode: c_int,
             ) -> c_int;
 
+            fn alarm(
+                seconds: c_uint,
+            ) -> c_uint;
+
+            fn atoi(
+                s: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+            ) -> c_int;
+
+            fn bind(
+                socket: c_int,
+                address: *const sockaddr: [READ],
+                address_len: socklen_t,
+            ) -> c_int;
+
+            fn calloc(
+                nobj: size_t,
+                size: size_t,
+            ) -> *mut c_void: [READ | WRITE | OFFSET_ADD | FREE];
```

Comment:
> I think the return type here should actually have `OFFSET_SUB` in addition to `OFFSET_ADD`. It's legal and reasonable to do things like `calloc(...).offset(1).offset(-1)`, and it seems difficult to specify something like "you can do `OFFSET_SUB` on this pointer, but only if you do an `OFFSET_ADD` first".

Why couldn't we have something like this for `offset`?

```rust
fn offset<T>(self: *const T: [OFFSET_ADD], count: isize) -> *const T: [OFFSET_ADD | OFFSET_SUB];
```

> (Though really `calloc` should probably be special-cased along with `malloc`, `realloc`, etc at some point.)

I think it already is.  I just added all of the `libc` `fn`s used by `lighttpd_rust_amalgamated` for consistency, even if they are already builtins.

Reply:
In dc4c467a, I made the `fn`s that internally do `.offset`s return `OFFSET_SUB`, which I think is clearly correct (e.x. for `strchr`).  I'm still not sure about the others, though.  From what I understood, if I pointer has only `OFFSET_ADD`, it can be made into a slice.  But if it has `OFFSET_SUB`, too, then it can't become a slice.  When a `fn` returns a pointer to the beginning of memory that could be a slice, why would we add `OFFSET_SUB`?